e2e/test/FaultInjection.cs

```diff
@@ -173,35 +174,75 @@ public static async Task ActivateFaultInjection(Client.TransportType transport,
-                int delay = FaultInjection.WaitForDisconnectMilliseconds - (int)watch.ElapsedMilliseconds;
-                if (delay < 0) delay = 0;
+                s_log.WriteLine($"{nameof(FaultInjection)}: Waiting for fault injection to be active: {delayInSec} seconds.");
+                await Task.Delay(TimeSpan.FromSeconds(delayInSec)).ConfigureAwait(false);
 
-                s_log.WriteLine($"{nameof(FaultInjection)}: Waiting for fault injection to be active: {delay}ms");
-                await Task.Delay(delay).ConfigureAwait(false);
+                // For disconnect type faults, the device should disconnect and recover.
+                if (FaultShouldDisconnect(faultType))
+                {
+                    s_log.WriteLine($"{nameof(FaultInjection)}: Confirming fault injection has been actived.");
+                    // Check that service issued the fault to the faulting device
+                    bool isFaulted = false;
+                    for (int i = 0; i < StatusCheckLoop; i++)
+                    {
+                        if (connectionStatusChangeCount >= 2)
+                        {
+                            isFaulted = true;
+                            break;
+                        }
 
-                await testOperation(deviceClient, testDevice).ConfigureAwait(false);
+                        await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
+                    }
+
+                    Assert.IsTrue(isFaulted, $"The device {testDevice.Id} did not get faulted with fault type: {faultType}");
+                    s_log.WriteLine($"{nameof(FaultInjection)}: Confirmed fault injection has been actived.");
+
+                    // Check the device is back online
+                    s_log.WriteLine($"{nameof(FaultInjection)}: Confirming device back online.");
+                    for (int i = 0; lastConnectionStatus != ConnectionStatus.Connected && i < durationInSec + StatusCheckLoop; i++)
+                    {
+                        await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false);
+                    }
+
+                    Assert.AreEqual(lastConnectionStatus, ConnectionStatus.Connected, $"{testDevice.Id} did not reconnect.");
+                    s_log.WriteLine($"{nameof(FaultInjection)}: Confirmed device back online.");
+
+                    // Perform the test operation.
+                    s_log.WriteLine($">>> {nameof(FaultInjection)}: Performing test operation for device {testDevice.Id}.");
+                    await testOperation(deviceClient, testDevice).ConfigureAwait(false);
+                }
+                else
+                {
+                    s_log.WriteLine($"{nameof(FaultInjection)}: Performing test operation while fault injection is being activated.");
+                    // Perform the test operation for the faulted device multi times.
+                    for (int i = 0; i < StatusCheckLoop; i++)
```

Comment:
There was inconsistent delay before fault injection occurs. And it did slightly increase execution time but it will make sure the test stable. I'll try to reduce it to 5 sec but not sure if it will consistent pass.